gtsfm/view_graph_estimator/view_graph_estimator_base.py

```diff
@@ -0,0 +1,123 @@
+"""Implements a base class for ViewGraph estimation.
+
+Estimating the ViewGraph can be done trivially by adding all the two-view estimates into a ViewGraph data structure.
+The purpose of this class, however, is to define an API for more sophisticated methods for estimating a ViewGraph 
+that include filtering or optimizing the two-view estimates.
+
+Authors: Akshay Krishnan, Ayush Baid
+"""
+import abc
+from typing import Dict, Optional, List, Tuple
+
+import dask
+import numpy as np
+from dask.delayed import Delayed
+from gtsam import Cal3Bundler, PinholeCameraCal3Bundler, Rot3, Unit3
+
+import gtsfm.utils.geometry_comparisons as comp_utils
+from gtsfm.evaluation.metrics import GtsfmMetric, GtsfmMetricsGroup
+from gtsfm.common.keypoints import Keypoints
+from gtsfm.common.view_graph import ViewGraph
+
+METRIC_GROUP = "view_graph"
+
+
+class ViewGraphEstimatorBase(metaclass=abc.ABCMeta):
+    """Base class for ViewGraph estimation.
+
+    A ViewGraphEstimator aggregates two-view estimates into a ViewGraph.
+    It could also improve the two-view estimates using filtering or optimization techniques.
+    """
+
+    @abc.abstractmethod
+    def run(
+        self,
+        i2Ri1: Dict[Tuple[int, int], Rot3],
+        i2Ui1: Dict[Tuple[int, int], Unit3],
+        calibrations: List[Cal3Bundler],
+        corr_idxs_i1i2: Dict[Tuple[int, int], np.ndarray],
+        keypoints: List[Keypoints],
+    ) -> ViewGraph:
```

Comment:
I think this should also return the filtered edges and vertices, and then this could be passed to `compute_metrics` to compute the metrics with respect to both the inliers and outliers as it's done in the `CycleConsistentRotationViewGraphEstimator`.

Reply:
Also, why not have `run` just take a `ViewGraph` as input? The same information that's being passed in is stored in the `Viewgraph` (except keypoints).

Reply:
I understand its not ideal to recompute them in compute_metrics(), but returning "inlier indices" assumes that the class only does filtering. It is designed to also support approaches that may refine the edges using optimization. 

about your second point, yeah I considered that. It will introduce some overhead in the user of this class to create a ViewGraph, and that seems like something this class should be doing.

Reply:
> I understand its not ideal to recompute them in compute_metrics(), but returning "inlier indices" assumes that the class only does filtering. It is designed to also support approaches that may refine the edges using optimization.

If no edges got filtered, then it simply wouldn't return a filtered graph; just make it Optional.

> about your second point, yeah I considered that. It will introduce some overhead in the user of this class to create a ViewGraph, and that seems like something this class should be doing.

What would the overhead be? You can just create a ViewGraph after running the TwoView Estimator on all the putative image pairs and then all the ViewGraphEstimators can operate off that.

Reply:
> If no edges got filtered, then it simply wouldn't return a filtered graph; it can Optional.

This is not functional style.

Reply:
> What would the overhead be? You can just create a ViewGraph after running the TwoView Estimator on all the putative image pairs and then all the ViewGraphEstimators can operate off that.

I am talking about just the two lines. Overlapping discussion in the other thread.

Reply:
I think just having the single line to create the initial ViewGraph would be much less overhead than having to deconstruct it into its constituent parts everytime we want to pass the output to another estimator IMHO